controllers/networking-calico/pkg/charts/utils.go

```diff
@@ -27,68 +30,168 @@ const (
-func ComputeCalicoChartValues(network *extensionsv1alpha1.Network, config *calicov1alpha1.NetworkConfig) map[string]interface{} {
-	var (
-		calicoChartValues = map[string]interface{}{
-			"images": map[string]interface{}{
-				calico.CNIImageName:                                        imagevector.CalicoCNIImage(),
-				calico.TyphaImageName:                                      imagevector.CalicoTyphaImage(),
-				calico.KubeControllersImageName:                            imagevector.CalicoKubeControllersImage(),
-				calico.NodeImageName:                                       imagevector.CalicoNodeImage(),
-				calico.PodToDaemonFlexVolumeDriverImageName:                imagevector.CalicoFlexVolumeDriverImage(),
-				calico.TyphaClusterProportionalAutoscalerImageName:         imagevector.TyphaClusterProportionalAutoscalerImage(),
-				calico.TyphaClusterProportionalVerticalAutoscalerImageName: imagevector.TyphaClusterProportionalVerticalAutoscalerImage(),
-			},
-			"global": map[string]string{
-				"podCIDR": network.Spec.PodCIDR,
-			},
-			"ipip": calicov1alpha1.Always,
-		}
-		calicoConfigValues = map[string]interface{}{
-			"backend": calicov1alpha1.Bird,
-		}
-		ipamConfig = map[string]interface{}{
-			"type":   hostLocal,
-			"subnet": usePodCIDR,
-		}
-		typhaConfig = map[string]interface{}{
-			"enabled": true,
-		}
-	)
+func ComputeCalicoChartValues(network *extensionsv1alpha1.Network, config *calicov1alpha1.NetworkConfig) (map[string]interface{}, error) {
+	typedConfig, err := generateChartValuesFrom(config)
+	if err != nil {
+		return nil, fmt.Errorf("error when generating calico config: %v", err)
+	}
+	calicoConfig, err := typedConfig.toMap()
+	if err != nil {
+		return nil, fmt.Errorf("could not convert calico config: %v", err)
+	}
+	calicoChartValues := map[string]interface{}{
+		"images": map[string]interface{}{
+			calico.CNIImageName:                                        imagevector.CalicoCNIImage(),
+			calico.TyphaImageName:                                      imagevector.CalicoTyphaImage(),
+			calico.KubeControllersImageName:                            imagevector.CalicoKubeControllersImage(),
+			calico.NodeImageName:                                       imagevector.CalicoNodeImage(),
+			calico.PodToDaemonFlexVolumeDriverImageName:                imagevector.CalicoFlexVolumeDriverImage(),
+			calico.TyphaClusterProportionalAutoscalerImageName:         imagevector.TyphaClusterProportionalAutoscalerImage(),
+			calico.TyphaClusterProportionalVerticalAutoscalerImageName: imagevector.TyphaClusterProportionalVerticalAutoscalerImage(),
+		},
+		"global": map[string]string{
+			"podCIDR": network.Spec.PodCIDR,
+		},
+		"config": calicoConfig,
+	}
+	return calicoChartValues, nil
+}
 
-	if config != nil {
-		switch config.Backend {
-		case calicov1alpha1.Bird, calicov1alpha1.None:
-			calicoConfigValues["backend"] = config.Backend
+func generateChartValuesFrom(config *calicov1alpha1.NetworkConfig) (*calicoConfig, error) {
```

Comment:
`generateChartValuesFrom` from what?
I think that only `generateChartValues` is good enough.